teps/0121-refine-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,201 @@
+|:---|:---|:---|:---|:---|:---|:---|
+| **When to Retry** | configurable | on failure |configurable|configurable|[always retry, conditional stop](https://github.com/ansible/ansible/pull/76101) [^ansible-conditional-stop]|configurable|
+| **Attempts amount** |supported|supported|supported|supported|supported|supported|
+| **Timeout for each attempt** |supported|supported|supported|[supported](https://docs.gitlab.com/ee/ci/yaml/#retrywhen)|supported|supported|
+| **Timeout for all attempts** |[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[?](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout)|-|-|-|
+
+Several observations regarding to the feature table above:
+- We can configure **timeout duration per attempt** in all CI systems that support the `retry` functionality.
+- GitHub Action doesn't support retry natively, but because the flexibility of **customized actions**, some users write their own `retry` action to make it work, and those customized actions even support what to do before retrying a failed attempt.
+- Concourse mentioned the reason it [retries per attempt is somewhat arbitrary](https://concourse-ci.org/attempts-step.html#attempts-step).
+- Argo Workflow supports timeout per attempt, but [this stackoverflow question](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout) implies the need of "timeout for all retries".
+
+## Options Under Consideration
+
+No matter how we implement the retry functionality, we propose to set `Timeout` for each retry attempt. This is decided based on the investigation about other CI/CD systems, see [related work](#related-work).
+
+### Option 1: Implement `retries` for Standalone TaskRun
+- Stop relying on `len(retriesStatus)` to determine whether a TaskRun or CustomRun finishes, use `ConditionSucceeded` & `ConditionFalse` instead.
+- `Retries` and `Timeout` are passed from `PipelineTask` to `TaskRunSpec` and `CustomRunSpec`.
+
+Three sub-options about the way to implement `retriesStatus`:
+- 1.a: Update `retriesStatus` for each retry attempt for `TaskRun`, keep `retriesStatus` for `CustomRun`
+  - No API change
+  - Need to implement a strategy for clients to get the previous pod and read its logs.
+- 1.b: Update `retriesStatus` for each retry attempt for `TaskRun`, deprecate `retriesStatus` for `CustomRun`
+  - No implementation restrictions of `retriesStatus` for `CustomRun`
+  - Need to implement a strategy for clients to get the previous pod and read its logs.
+- 1.c: Deprecate `retriesStatus` for both `TaskRun` and `CustomRun`, create a new `TaskRun` for each retry attempt.
```

Comment:
For independent `taskRun`, the users might have to query metadata (labels and annotations) to get the list of `taskRun` CRDs for each retry attempt. Let's say there is a `taskRun` which was retried five times, each five `taskRuns` and `pod` will have a label saying it was created for a particular instance.

For the `taskRun` created for a `pipelineTask`, how do we retrieve the collection of `taskRuns` for each attempt? Using the same metadata?

`taskRun` with `retriesStatus` works as a master with a list of children as each attempt. It hides complexity of querying a list of taskRuns for a given `pipelineTask` and checking the status of each attempted `taskRun`.

Reply:
Summary our recent dicussion offline around option `1.c`, the details are attached in Appendix if you are interested.

1. TaskRun reconciler creates a new `TaskRun` for each retry attempt.
2. Creates a new field `RetryAttempts` in `TaskRun.Status` to record the retry history.
3. Creates two new labels:
    - `tekton.dev/retry-count: <retry number>`, if the number is 0, the TaskRun is not initiated for a retry.
    - `tekton.dev/retry-parent: <parent taskrun name>`, record the parent TaskRun of a retry attempt TaskRun.

We need to prototype this solution before moving this TEP to implementable.

Thanks for all of the inputs and suggestions @pritidesai !